src/rtc.rs

```diff
@@ -0,0 +1,124 @@
+//! Real Time Clock
+
+use stm32f103xx::{RCC, PWR, RTC};
+
+/// RTC clock source
+#[derive(Clone, Copy, Debug)]
+pub enum RtcClkSource {
+    /// No clock
+    NoClock = 0b00,
+    /// LSE oscillator clock (32.768 kHz), 
+    LSE = 0b01, 
+    /// LSI oscillator clock (~ 40kHz)
+    LSI = 0b10,
+    /// HSE oscillator clock (4-16 MHz) divided by 128 
+    HSE = 0b11,
+}
+
+/// RTC interrupt event
+pub enum RtcEvent {
+    /// "Second" Tick Interrupt
+    Second,
+    /// Alarm
+    Alarm,
+    /// Overflow
+    Overflow,
+}
+
+/// RTC
+pub struct Rtc<'a>(pub &'a RTC);
+
+impl<'a> Rtc<'a> {
+    /// Initializes RTC
+    pub fn init(&self, clk_source: RtcClkSource, prescalar: u32, rcc: &RCC, pwr: &PWR) {
+        let rtc = self.0;
+
+        // Enable backup and power interface clocks
+        rcc.apb1enr.modify(|_, w| w.bkpen().enabled().pwren().enabled());
+
+        // unprotect backup and RTC registers
+        pwr.cr.modify(|_, w| w.dbp().set_bit());
+
+        // Reset backup domain to allow changing clock source
+        rcc.bdcr.modify(|_, w| w.bdrst().set_bit());
+        rcc.bdcr.modify(|_, w| w.bdrst().clear_bit());
+
+        match clk_source {
+            // Turn on clock source and wait for it to be ready
+            RtcClkSource::LSE => {
+                rcc.bdcr.modify(|_, w| w.lseon().set_bit());
+                while rcc.bdcr.read().lserdy().bit_is_clear() { };
+            },
+            RtcClkSource::LSI => {
+                rcc.csr.modify(|_, w| w.lsion().set_bit());
+                while rcc.csr.read().lsirdy().bit_is_clear() { };
+            },
+            RtcClkSource::HSE => {
+                rcc.cr.modify(|_, w| w.hseon().set_bit());
+                while rcc.cr.read().hserdy().bit_is_clear() { };
+            },
+            RtcClkSource::NoClock => { }
```

Comment:
I'm curious about the use case for this variant. It seems that doing nothing here will make the program use the previous configuration, maybe even the reset value configuration. It seems that by selecting `NoClock` the user has no guarantee about how the clock will be configured at the end (but may I'm missing something here).

Reply:
I am not quite sure on the use case of NoClock myself. I added it since it was noted as a possible value of the rtc clock selection register bits in the chip spec (as well as making the enum values work out before I hardcoded them :P).

This match block is purely for turning on the clock. Immediately below is where the rtc clock selection register bits are actually set - so even in the case of NoClock they are written.
```rcc.bdcr.modify(|_, w| unsafe { w.rtcsel().bits(clk_source as u8) });```

Reply:
Ohh, OK. Missed that part. Never mind then.